feat: additional validation and exception handler

bezzang-dev opened the pull request:
This PR added Validation and ExceptionHandler.

- Added validation annotations to each DTO field that requires validation.
- Added validation annotations to each parameter that requires validation in each controller.
- Added exception handling methods to the GlobalAPIExceptionHandler.

This PR closes #77.

bezzang-dev committed:
Feature: Implement deleteAnswer

bezzang-dev committed:
Feature: Add annotations for fields that require validation in DTO

bezzang-dev committed:
Feature: Add annotations for methods and parameters that require validation check in Controller class

bezzang-dev committed:
Fix: Resolved the issue where BaseTimeEntity was not being used

bezzang-dev committed:
Feature: implement to method for the GlobalAPIExceptionHandler

bezzang-dev committed:
Feature: Add @NotBlank in title, description

bezzang-dev committed:
Fix: delete @NotBlank because UUID and List type are not supported

seiwonpark commented on .github/workflows/gradle.yml:
```diff
@@ -2,9 +2,9 @@ name: SpringBoot CI with Gradle
 on:
   push:
-    branches: ["develop", "feature/validate-backend"]
+    branches: ["develop", "feature/validator"]
```
Please don't update build file multiple times. It makes hard tracking bugs or errors especially like this PR which contains lots of file is changed.

You can use `feature/**` to match all the branches for this

seiwonpark commented on api/src/main/java/com/thesurvey/api/controller/AuthenticationController.java:
```diff
@@ -19,6 +18,7 @@
 import io.swagger.annotations.ApiResponses;
 
 @Api(tags = "인증")
+@Validated
```
What is difference using @Validated and @Valid annotation for each controllers?

bezzang-dev replied:
@Validated annotation can specify groups for validation of methods or classes. and This performs validation prior to method execution. If validation fails, a MethodArgumentNotValidException exception occurs.

@Valid annotation, on the other hand, performs validation of the fields in the object. It performs validation at the time when you create the parameter object. If validation fails, a ConstructViolationException exception is generated.

but since we don't use validation groups, I will revise whether these two annotations need to be used together.

seiwonpark replied:
Thank you for this explanation!

seiwonpark commented on api/src/main/java/com/thesurvey/api/controller/AuthenticationController.java:
```diff
@@ -55,12 +50,9 @@ public ResponseEntity<UserResponseDto> register(
-        } catch (AuthenticationException e) {
-            throw new ExceptionMapper(ErrorMessage.UNAUTHORIZED_REQUEST);
-        }
+    public ResponseEntity<?> login(
```
As we're using `request` and `response` types, we should specify the object type.

seiwonpark commented on api/src/main/java/com/thesurvey/api/controller/SurveyController.java:
```diff
@@ -78,9 +80,9 @@ public ResponseEntity<SurveyResponseDto> getSurvey(
-    public ResponseEntity<SurveyResponseDto> createSurvey(Authentication authentication,
-        @RequestBody SurveyRequestDto surveyRequestDto) {
-        return ResponseEntity.ok(surveyService.createSurvey(authentication, surveyRequestDto));
+    public ResponseEntity<?> createSurvey(Authentication authentication,
```
Likewise, as we've already specified response type(DTO), this should be specified.

seiwonpark commented on api/src/main/java/com/thesurvey/api/dto/request/AnsweredQuestionRequestDto.java:
```diff
@@ -11,6 +12,7 @@ public class AnsweredQuestionRequestDto {
 
     private UUID surveyId;
 
+    @NotBlank
```
Same reason for https://github.com/kookmin-sw/capstone-2023-40/pull/78/commits/50685a2ec3d483ef55d2f8c2262fba21684a9f99

seiwonpark commented on api/src/main/java/com/thesurvey/api/exception/GlobalAPIExceptionHandler.java:
```diff
@@ -1,16 +1,65 @@
-        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
+    /*
+        @handleExceptionMapperException handles the exception thrown by ExceptionMapper.
+    */
```
It doesn't apply to JavaDoc. Please modify this to 

```java
/**
 * Handles exception thrown by ExceptionMapper.
 */
```

And please I'm repeating this: **Follow code conventions**.

seiwonpark commented on api/src/main/java/com/thesurvey/api/exception/GlobalAPIExceptionHandler.java:
```diff
@@ -1,16 +1,65 @@
+
+    /*
+        @handleValidationException handles the exception thrown by the validation of the DTO.
+    */
```
Likewise modify all your comments as above.

seiwonpark commented on api/src/main/java/com/thesurvey/api/exception/GlobalAPIExceptionHandler.java:
```diff
@@ -1,16 +1,65 @@
+            .findFirst()
+            .orElse(new ExceptionMapper(ErrorMessage.INVALID_REQUEST).getMessage());
+        return ResponseEntity.badRequest().body(errorMessage);
+    }
```
This is very good!

seiwonpark reviewed (approved):
Please update your code and re-request.

bezzang-dev committed:
Refactor: revise valid annotations
- Added error message for each annotation
- Fixed inappropriate annotations
- Added logic to check if the survey was answered in duplicate by same user
- Added logic to check if the startedTime and endedTime were reversed

bezzang-dev committed:
Refactor: revise BaseTimeEntity and Controllers
- Modified to specify return type in controller
- Fix BaseTimeEntity used incorrectly

seiwonpark commented on api/src/main/java/com/thesurvey/api/dto/request/AnsweredQuestionRequestDto.java:
```diff
@@ -2,17 +2,18 @@
 @Builder
 public class AnsweredQuestionRequestDto {
 
+    @NotNull(message = "유효하지 않은 설문조사입니다.")
```
`message` is good but I think we can catch this error on `GlobalAPIExceptionHandler` which handles error on `non-property` or `null`. 

Well if this property is **required** to be not null or empty string(which is the most strong rule), then you should use `@NotBlank`.

- `@NotBlank`
  - checks empty string (ex. `""`)
  - checks null (ex. `null`)
  - checks blank (ex. `" "`)
- `@NotNull`
  - checks null (ex. `null`)
- `@NotEmpty`
  - checks empty string (ex. `""`)
  - checks blank (ex. `" "`)

So, my opinion overall is to remove `message` since there're `GlobalAPIExceptionHandler` which returns error message, and modify all `required porperties` to `@NotBlank` since it'll catch all 3 conditions.

But which do you think is the right way to do this?

bezzang-dev replied:
I wrote different error messages for each property validation to provide specific error messages,
 but if it's not necessary, I agree that We remove `message`. Or does it mean that we should write it in `GlobalAPIExceptionHandler`??

Upon further research, `ConstraintViolationException` is part of the Java Bean Validation API and is automatically handled by the Spring Boot validation framework, so it seems that the handleValidationException method can be removed.


~~So, I suggest removing `handleValidationException` method and keeping `message`.~~

I'll think about which one is better...
 
And `@NotBlank` annotation only supports `String` type, so it cannot be applied to `Numeric` or `List` types. 
Instead, I will modify it to use both `@NotNull` and `@NotEmpty` annotations.

bezzang-dev replied:
As you say, I think handling exceptions in `GlobalAPIExceptionHandler` is better. 

While defining messages for each property allows for more specific error messages, it can be cumbersome to write messages for each new property that is added. 
With `GlobalAPIExceptionHandler`, we can respond with error messages that are understandable to the client and allow for flexible exception handling. 

Thank you for your opinion!

seiwonpark replied:
When I've tested with your code(I'm not 100% for sure), without setting error messages, I could get message from `GlobalAPIExceptionHandler`. So we might need additional messages.

Anyway utilize `Global****` things!

seiwonpark commented on api/src/main/java/com/thesurvey/api/dto/request/UserRegisterRequestDto.java:
```diff
@@ -14,24 +15,24 @@
 
-    @NotBlank
+    @NotBlank(message = "비밀번호를 확인해주세요.")
+    @Min(value = 8, message = "비밀번호는 최소 8자 이상이어야 합니다.")
```
And also please add `@Max` to `25`.

bezzang-dev replied:
Oh.... I apologize for not fixing the issue of using the @Max and @Min annotations on a String type, which only support numeric types....  I will correct it by using the @Size annotation instead.. ;(

seiwonpark replied:
Very nice! Thank you!

seiwonpark commented on api/src/main/java/com/thesurvey/api/exception/GlobalAPIExceptionHandler.java:
```diff
@@ -12,20 +12,22 @@
     @ExceptionHandler(ExceptionMapper.class)
-    public ResponseEntity<String> handleExceptionMapperException(Exception e) {
-        return ResponseEntity.badRequest().body(e.getMessage());
+    public ResponseEntity<String> handleExceptionMapperException(Exception error) {
```
Thank you for this modification `e` to `error`.

I've mentioned and emphasized on **code conventions**. Code conventions is actually for cooperation since all other language, country, whatever people could understand beyond time. And it's really good to keep convention with **FULL NAME** because there's no way to misunderstand this.

But there're some agreeable conventions like `btn` for `button`, `tmp` for `temp`, `bg` for `background` like that. And this `e` is kind of one thing that is agreeable convention. And if you see some utils libraries, you'll see some abbreviated version of words like
```typescript
/**
 * Check if input text is empty string.
 *
 * @param {string} s String type text
 * @returns {boolean} Result of checking empty string
 */
export const isEmptyString = (s: string): boolean => {
  return s === '';
};
```
this way.

But anyway, this kind of modification is sooooo much better than any conventions. Very nice!

seiwonpark reviewed (approved):
Partial LGTM. You can merge this right away or after minor fixes.

bezzang-dev committed:
Refactor: Remove message, Add some notation

bezzang-dev committed:
Feature: Implement handleHttpMessageNotReadableException method

seiwonpark reviewed (approved):
LGTM. Thanks!
